module/zfs/zfs_ioctl.c

```diff
@@ -219,6 +232,34 @@ typedef int zfs_ioc_legacy_func_t(zfs_cmd_t *);
 typedef int zfs_ioc_func_t(const char *, nvlist_t *, nvlist_t *);
 typedef int zfs_secpolicy_func_t(zfs_cmd_t *, nvlist_t *, cred_t *);
 
+/*
+ * IOC Keys are used to document and validate user->kernel interface inputs.
+ * See zfs_keys_recv_new for an example declaration. Any key name that is not
+ * listed will be rejected as input.
+ *
+ * The keyname 'optional' is always allowed, and must be an nvlist if present.
+ * Arguments which older kernels can safely ignore can be placed under the
+ * "optional" key.
+ *
+ * When adding new keys to an existing ioc for new functionality, consider:
+ * 	- adding an entry into zfs_sysfs.c zfs_features[] list
+ * 	- updating the libzfs_input_check.c test utility
+ */
+typedef enum {
+	ZK_OPTIONAL = 1 << 0,		/* pair is optional */
+	ZK_WILDCARDLIST = 1 << 1,	/* name of bookmark|snapshot|property */
+					/* at least one name is required */
```

Comment:
Seems like WILKDCARDLIST is a pretty special case.

Are we checking that the names are valid bookmarks/snapshots/properties?  If not we probably shouldn't document it as such.  Instead we should say that any key name is accepted.

Seems like it would be hard to use WILKDCARDLIST with additional keys.

Maybe document that zkey_name isn't used for checking but is useful for documentation?